scapy/contrib/automotive/ccp.py

```diff
@@ -0,0 +1,583 @@
+    fields_desc = [
+        ByteField('data_length', 0),
+        ByteField('data_type', 0),
+        StrFixedLenField('ccp_reserved', b'\xff' * 3, length=3)
+    ]
+
+
+class ACTION_SERVICE_DTO(Packet):
+    fields_desc = [
+        ByteField('data_length', 0),
+        ByteField('data_type', 0),
+        StrFixedLenField('ccp_reserved', b'\xff' * 3, length=3)
+    ]
+
+
+class DTO(Packet):
+    __slots__ = Packet.__slots__ + ["payload_cls"]
+
+    return_codes = {
+        0x00: "acknowledge / no error",
+        0x01: "DAQ processor overload",
+        0x10: "command processor busy",
+        0x11: "DAQ processor busy",
+        0x12: "internal timeout",
+        0x18: "key request",
+        0x19: "session status request",
+        0x20: "cold start request",
+        0x21: "cal. data init. request",
+        0x22: "DAQ list init. request",
+        0x23: "code update request",
+        0x30: "unknown command",
+        0x31: "command syntax",
+        0x32: "parameter(s) out of range",
+        0x33: "access denied",
+        0x34: "overload",
+        0x35: "access locked",
+        0x36: "resource/function not available"
+    }
+    fields_desc = [
+        XByteField("packet_id", 0xff),
+        ByteEnumField('return_code', 0x00, return_codes),
+        ByteField('ctr', 0)
+    ]
+
+    def __init__(self, *args, **kwargs):
+        self.payload_cls = DEFAULT_DTO
+        if "payload_cls" in kwargs:
+            self.payload_cls = kwargs["payload_cls"]
+            del kwargs["payload_cls"]
+        Packet.__init__(self, *args, **kwargs)
+
+    def guess_payload_class(self, payload):
+        return self.payload_cls
+
+    @staticmethod
+    def get_dto_cls(cmd):
+        dto_cls_mapping = {
+            0x03: DNLOAD_DTO,
+            0x04: UPLOAD_DTO,
+            0x09: GET_ACTIVE_CAL_PAGE_DTO,
+            0x0D: GET_S_STATUS_DTO,
+            0x0E: BUILD_CHKSUM_DTO,
+            0x0F: SHORT_UP_DTO,
+            0x12: GET_SEED_DTO,
+            0x13: UNLOCK_DTO,
+            0x14: GET_DAQ_SIZE_DTO,
+            0x17: EXCHANGE_ID_DTO,
+            0x18: PROGRAM_DTO,
+            0x1B: GET_CCP_VERSION_DTO,
+            0x20: DIAG_SERVICE_DTO,
+            0x21: ACTION_SERVICE_DTO,
+            0x22: PROGRAM_6_DTO,
+            0x23: DNLOAD_6_DTO
+        }
+        return dto_cls_mapping[cmd] if cmd in dto_cls_mapping.keys() \
+            else DEFAULT_DTO
+
+    def answers(self, other):
+        """DEV: true if self is an answer from other"""
+        if not hasattr(other, "ctr"):
+            return 0
+        if self.ctr != other.ctr:
+            return 0
+        if not hasattr(other, "cmd"):
+            return 0
+
+        new_pl_cls = self.get_dto_cls(other.cmd)
+        if self.payload_cls != new_pl_cls and \
+                self.payload_cls == DEFAULT_DTO:
+            data = bytes(self.load)
+            self.remove_payload()
+            self.add_payload(new_pl_cls(data))
+            self.payload_cls = new_pl_cls
```

Comment:
Could you document a bit what this is doing ?

I don’t think one would expect `answers()` to edit the packets, it is commonly only a test

Reply:
Done